language-plutus-core/src/Language/PlutusCore/Analysis/Definitions.hs

```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE LambdaCase #-}
+-- | Definition analysis for Plutus Core.
+module Language.PlutusCore.Analysis.Definitions (UniqueInfos, ScopeType, termDefs, typeDefs, runTermDefs, runTypeDefs) where
+
+import           Language.PlutusCore.Error
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Type
+
+import           Data.Functor.Foldable
+
+import           Control.Lens              hiding (use, uses)
+import           Control.Monad.Except
+import           Control.Monad.State
+import           Control.Monad.Writer
+
+import           Data.Coerce
+import           Data.Foldable
+import qualified Data.IntMap               as IM
+import qualified Data.Set                  as Set
+
+{- Note [Unique usage errors]
+The definitions analysis can find a number of problems with usage of uniques, however
+clients may not always want to throw an error on all of them, hence we simply return
+them all and allow the client to chose if they want to throw some of them.
+-}
+
+-- | Information about a unique, a pair of a definition if we have one and a set of uses.
+type UniqueInfo a = (Maybe (ScopedLoc a), Set.Set (ScopedLoc a))
+type UniqueInfos a = IM.IntMap (UniqueInfo a)
```

Comment:
It's probably better to keep two maps - one for types and one for terms. Do you have any benchmarks of the performance of the approach used here (compared to the approach mentioned)?

Reply:
Well, part of the point of this is to detect uniques that are defined/used incoherently in both scopes, so keeping everything in the one map is convenient. I could pull it out but I don't think it would make anything clearer?

I can't imagine that's going to affect performance much, at most we would have two smaller maps instead of one bigger map, and we'd need to look up in both anyway.